File: python/camera_ui_sdk/sensor/siren.py
from __future__ import annotations

from collections.abc import Mapping
from enum import StrEnum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType


class SirenCapability(StrEnum):
    """Optional capabilities of a siren control."""

    Volume = "volume"
    """Siren supports volume adjustment (0-100)."""


class SirenProperty(StrEnum):
    """Property names of a siren control."""

    Active = "active"
    """Whether the siren is currently active."""
    Volume = "volume"
    """Volume level (0-100)."""


class SirenControlProperties(TypedDict):
    """Property values of a siren control."""

    active: bool
    volume: int


class SirenPropertyChangeData(TypedDict):
    """Property change payload emitted on SirenControlLike.onPropertyChanged."""

    property: str
    """Name of the changed property, a SirenProperty value."""
    value: bool | int
    """New value of the property."""


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class SirenControlLike(SensorLike, Protocol):
    """Read-only proxy interface for a siren control."""

    @property
    def type(self) -> SensorType:
        return SensorType.Siren

    @property
    def onPropertyChanged(self) -> Observable[SirenPropertyChangeData]: ...

    @property
    def onCapabilitiesChanged(self) -> Observable[list[SirenCapability]]: ...

    @overload
    def getValue(self, property: Literal[SirenProperty.Active]) -> bool | None: ...
    @overload
    def getValue(self, property: Literal[SirenProperty.Volume]) -> int | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...


class SirenControl(Sensor[SirenControlProperties, TStorage, SirenCapability], Generic[TStorage]):
    """Siren control sensor. Override `setActive()`/`setInactive()` to drive
    your hardware, then `await super().setActive()` / `await super().setInactive()`
    to sync the SDK state. For hardware-pushed updates, call the super methods
    from your event handler. That bypasses any plugin override and only syncs
    state.
    """

    _requires_frames = False

    def __init__(self, name: str = "Siren", *, native_id: str | None = None) -> None:
        super().__init__(name, native_id=native_id)
        self._write_state(
            {
                SirenProperty.Active.value: False,
                SirenProperty.Volume.value: 100,
            }
        )

    @property
    def type(self) -> SensorType:
        return SensorType.Siren

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Control

    @property
    def active(self) -> bool:
        return bool(self.props.active)

    @property
    def volume(self) -> int:
        value = self.props.volume
        return int(value) if value is not None else 0

    async def setActive(self) -> None:
        """Activate the siren. Override to drive hardware and call
        `await super().setActive()` after success to sync the SDK state.

        Example:
            ```python
            await siren.setActive()
            ```
        """
        self._write_state({SirenProperty.Active.value: True})

    async def setInactive(self) -> None:
        """Deactivate the siren. Override to drive hardware and call
        `await super().setInactive()` after success to sync the SDK state.

        Example:
            ```python
            await siren.setInactive()
            ```
        """
        self._write_state({SirenProperty.Active.value: False})

    async def setVolume(self, value: int) -> None:
        """Set volume. Override to drive hardware and call
        `await super().setVolume(value)` after success. The default implementation
        clamps the value to [0, 100].

        Args:
            value: Volume level in the range 0-100.

        Example:
            ```python
            await siren.setVolume(80)
            ```
        """
        self._write_state({SirenProperty.Volume.value: max(0, min(100, value))})

    async def updateValue(self, property: str, value: Any) -> None:
        """Routes generic property writes to the semantic setters."""
        if property == SirenProperty.Active.value:
            if value:
                await self.setActive()
            else:
                await self.setInactive()
            return
        if property == SirenProperty.Volume.value:
            await self.setVolume(int(value))
            return

File: node/src/sensor/siren.ts
import { Sensor, SensorType, SensorCategory } from './base.js';
import { defineSensor, SensorDomain } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike, SensorOptions } from './base.js';

/** Optional capabilities of a siren control. */
export enum SirenCapability {
  /** Siren supports volume adjustment (0-100). */
  Volume = 'volume',
}

/**
 * Property names of a siren control.
 *
 * @internal
 */
export enum SirenProperty {
  /** Whether the siren is currently active. */
  Active = 'active',
  /** Volume level (0-100). */
  Volume = 'volume',
}

/**
 * Property values of a siren control.
 *
 * @internal
 */
export interface SirenControlProperties {
  [SirenProperty.Active]: boolean;
  [SirenProperty.Volume]: number;
}

/** Read-only proxy interface for a siren control. */
export interface SirenControlLike extends SensorLike {
  readonly type: SensorType.Siren;
  readonly onPropertyChanged: Observable<PropertyChangeOf<SirenControlProperties>>;
  readonly onCapabilitiesChanged: Observable<SirenCapability[]>;

  getValue(property: SirenProperty.Active): boolean | undefined;
  getValue(property: SirenProperty.Volume): number | undefined;
  getValue(property: string): unknown;
}

/**
 * Siren control sensor. Override `setActive()`/`setInactive()` to drive your
 * hardware, then `await super.setActive()` / `await super.setInactive()` to sync
 * the SDK state. For hardware-pushed updates, call the `super` methods from your
 * event handler. That bypasses any plugin override and only syncs state.
 */
export class SirenControl<TStorage extends object = Record<string, any>> extends Sensor<SirenControlProperties, TStorage, SirenCapability> {
  readonly type = SensorType.Siren;
  readonly category = SensorCategory.Control;

  constructor(name = 'Siren', options?: SensorOptions) {
    super(name, options);

    this._writeState({
      [SirenProperty.Active]: false,
      [SirenProperty.Volume]: 100,
    });
  }

  get active(): boolean {
    return this.props.active;
  }

  get volume(): number {
    return this.props.volume;
  }

  /**
   * Activate the siren. Override to drive hardware and call `await super.setActive()`
   * after success to sync the SDK state.
   *
   * @example
   * ```ts
   * await siren.setActive();
   * ```
   */
  async setActive(): Promise<void> {
    this._writeState({ [SirenProperty.Active]: true });
  }

  /**
   * Deactivate the siren. Override to drive hardware and call `await super.setInactive()`
   * after success to sync the SDK state.
   *
   * @example
   * ```ts
   * await siren.setInactive();
   * ```
   */
  async setInactive(): Promise<void> {
    this._writeState({ [SirenProperty.Active]: false });
  }

  /**
   * Set volume. Override to drive hardware and call `await super.setVolume(value)`
   * after success. The default implementation clamps the value to [0, 100].
   *
   * @param value - Volume level in the range 0-100.
   *
   * @example
   * ```ts
   * await siren.setVolume(80);
   * ```
   */
  async setVolume(value: number): Promise<void> {
    this._writeState({ [SirenProperty.Volume]: Math.max(0, Math.min(100, value)) });
  }

  /**
   * Routes generic property writes to the semantic setters.
   *
   * @internal
   */
  override async updateValue(property: string, value: unknown): Promise<void> {
    switch (property as SirenProperty) {
      case SirenProperty.Active:
        if (value) await this.setActive();
        else await this.setInactive();
        return;
      case SirenProperty.Volume:
        await this.setVolume(value as number);
        return;
    }
  }
}

/** Registry metadata for {@link SirenControl}. */
export const sirenMeta = defineSensor({
  type: SensorType.Siren,
  category: SensorCategory.Control,
  assignmentKey: 'siren',
  multiProvider: true,
  isDetectionType: false,
  properties: {
    [SirenProperty.Active]: { type: 'boolean', writable: true },
    [SirenProperty.Volume]: { type: 'number', min: 0, max: 100, unit: '%', writable: true },
  },
  shortcutable: true,
  cascadeTrigger: { property: SirenProperty.Active, value: true, sustained: true },
  propertyCapabilities: { [SirenProperty.Volume]: SirenCapability.Volume },
  virtual: { properties: { [SirenProperty.Active]: false, [SirenProperty.Volume]: 100 }, capabilities: [SirenCapability.Volume] },
  semantics: {
    domain: SensorDomain.Siren,
    stateProperty: SirenProperty.Active,
    commandProperty: SirenProperty.Active,
  },
});
